compiler/AST/foralls.cpp

```diff
@@ -582,9 +591,146 @@ static bool findStandaloneOrLeader(ForallStmt* pfs, CallExpr* iterCall,
       USR_FATAL(iterCall, "A%s leader iterator is not found for the iterable expression in this forall loop", pfs->zippered() ? "" : " standalone or");
   }
 
-  return gotSA;
+  return retval;
+}
+
+/////////// handleZipperedSerial /////////// 
+
+static FnSymbol* trivialLeader          = NULL;
+static Type*     trivialLeaderYieldType = NULL;
+
+static void hzsCheckParallelIterator(ForallStmt* fs, FnSymbol* origIterFn) {
+  if (isLeaderIterator(origIterFn) || isStandaloneIterator(origIterFn)) {
+    USR_FATAL_CONT(fs->iteratedExpressions().head,
+          "cannot use a parallel iterator as the first iterable expression"
+          "in a zippered forall loop");
+    USR_PRINT(origIterFn, "the parallel iterator being invoked is here");
```

Comment:
I'm really confused by this error message. Can you create a test that runs into this error message? Isn't this code only run if the first iterator is serial? And anyway, we zipper parallel iterators in forall loops all the time. What's this error really about? Is there some case that's not implemented yet?

Reply:
Good question! I looked more carefully and right now I think it is not possible to hit this case.

Here is the closest we can get:
```chpl
for idx in zip(myIter(tag=standalone), anotherIter()) do .....;
```

Right now we handle this case as a (zippered) serial loop. (And, ironically, generate invalid C code.) However, if we are to handle it, it needs to be handled as a forall if there is an enclosing forall loop with, say, reduce intents:

```chpl
forall .... with (+ reduce sum) {
  for idx in zip(myIter(tag=standalone), anotherIter()) do
    sum reduce= f(idx);
}
```

If the compiler handles the inner loop as a `for` loop, then it does not create shadow variables for `sum` for the various tasks that `myIter(tag=standalone)` may be creating. So all those tasks will have a data race accessing the same (shadow) variable for `sum`.

So, I am going to make a separate PR to issue a user error "sorry this case is not implemented because it can potentially lead to data races" for the zippered for-loop above. Maybe I will even have the compiler handle the above case as a forall loop, which should trigger the error that is not triggered now.

Popping back, right now we cannot hit that error `cannot use a parallel iterator as the first iterable .....`. However, I would like to retain that check to prevent silently generating racy code if something changes in the future and the user manages to write code that triggers it.

I could convert that user error to INT_ASSERT. Pro: if the user gets it, they can report it so we will hopefully know about it. Cons: the user will miss an explanation of what is going on and will not be able to adjust their code without hearing from us. So I propose keeping it a user error.

Reply:
For this PR I'd be pretty happy if you simply changed the error to say something like "Support for this combination of zippered iterators is not currently implemented"